packages/api/src/external/hie/register-facility.ts.ts

```diff
@@ -0,0 +1,110 @@
+import {
+  AddressStrict,
+  AddressWithCoordinates,
+  removeCoordinates,
+} from "@metriport/core/domain/location-address";
+import {
+  FacilityRegister,
+  Facility,
+  FacilityCreate,
+  isOboFacility,
+} from "../../domain/medical/facility";
+import { buildCqOrgName } from "../../external/carequality/shared";
+import { buildCwOrgName } from "../../external/commonwell/shared";
+import { getCxOrganizationNameOidAndType } from "../../command/medical/organization/get-organization";
+import { getAddressWithCoordinates } from "../../domain/medical/address";
+import { getCqOboData } from "../../external/carequality/get-obo-data";
+import { createOrUpdateFacility } from "../../command/medical/facility/create-or-update-facility";
+import { createOrUpdateInCq, createOrUpdateInCw } from "./shared";
+
+/**
+ * Registers a new facility within HIEs.
+ *
+ * @param cxId
+ * @param facility
+ * @returns The updated facility.
+ */
+export async function registerFacilityWithinHIEs(
+  cxId: string,
+  facility: FacilityRegister
+): Promise<Facility> {
+  const isObo = isOboFacility(facility.type);
+
+  const [cxOrg, address, cqOboData] = await Promise.all([
+    getCxOrganizationNameOidAndType(cxId),
+    getAddressWithCoordinates(getAddressFromInput(facility), cxId),
+    getCqOboData(facility.cqOboActive, facility.cqOboOid),
+  ]);
+
+  const facilityDetails = createFacilityDetails(cxId, facility, address);
+
+  let oboFacilityDetails = facilityDetails;
+
+  if (isObo) {
+    oboFacilityDetails = {
+      ...oboFacilityDetails,
+      cqOboActive: facility.cqOboActive,
+      cqOboOid: facility.cqOboOid,
+      cwOboActive: facility.cwOboActive,
+      cwOboOid: facility.cwOboOid,
+    };
+  }
+
+  const cmdFacility = await createOrUpdateFacility(
+    cxId,
+    facility.id,
+    facility.data.npi,
+    oboFacilityDetails
+  );
+
+  // CAREQUALITY
+  const cqOboEnabled = isObo && cqOboData.enabled;
+  if (cqOboEnabled || !isObo) {
+    const cqFacilityName = cqOboEnabled ? cqOboData.cqFacilityName : cmdFacility.data.name;
+    const cqOboOid = cqOboEnabled ? cqOboData.cqOboOid : undefined;
+    const cqOrgName = buildCqOrgName(cxOrg.name, cqFacilityName, cqOboOid);
+    await createOrUpdateInCq(cmdFacility, cxOrg.oid, cqOrgName, address);
+  }
+
+  // COMMONWELL
+  const cwOboEnabled = isObo && facility.cwOboActive && facility.cwOboOid;
+  if (cwOboEnabled || !isObo) {
+    const cwFacilityName = facility.cwFacilityName ?? cmdFacility.data.name;
+    const cwOrgName = buildCwOrgName(cxOrg.name, cwFacilityName, facility.cwOboOid);
```

Comment:
It make a request w/ these params we'd end up with the wrong Org name:
- facility type: `initiator_and_responder`
- cwOboOid: set to whatever

It will end up w/ `-OBO-` on the name.

We should prob send `isObo` as param as well.

Reply:
I used isProvider